<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1850]:
import pandas as pd
import psycopg2

In [1851]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [1852]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [1853]:
# текст запроса
query_3_1 = '''
SELECT COUNT(*) AS total_vacancies
FROM public.vacancies
'''

In [1854]:
# результат запроса
df = pd.read_sql(query_3_1, connection)

print(df.iloc[0]['total_vacancies'])

49197


/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/233460687.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query_3_1, connection)


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [1855]:
# текст запроса
query_3_2 = '''
SELECT COUNT(*) AS total_employers
FROM public.employers
'''

In [1856]:
# результат запроса
df = pd.read_sql(query_3_2, connection)
print(df.iloc[0]['total_employers'])

/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/1383761665.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query_3_2, connection)


23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [1857]:
# текст запроса
query_3_3 = '''
SELECT COUNT(*) AS total_areas
FROM public.areas
'''

In [1858]:
# результат запроса
df = pd.read_sql(query_3_3, connection)
print(df.iloc[0]['total_areas'])

1362


/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/2243038636.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query_3_3, connection)


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [1859]:
# текст запроса
query_3_4 = '''
SELECT COUNT(*) AS total_industries
FROM public.industries
'''

In [1860]:
# результат запроса
df = pd.read_sql(query_3_4, connection)
print(df.iloc[0]['total_industries'])

294


/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/117896274.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query_3_4, connection)


***

In [1861]:
# выводы по предварительному анализу данных

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [1862]:
# текст запроса
query_4_1 = '''
SELECT 
    a.name AS area,
    COUNT(v.id) AS cnt
FROM public.vacancies v
JOIN public.areas a ON v.area_id = a.id
GROUP BY a.id, a.name
ORDER BY cnt DESC
LIMIT 5
'''

In [1863]:
# результат запроса
df = pd.read_sql(query_4_1, connection)
print(df)

              area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892


/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/492581935.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query_4_1, connection)


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [1864]:
# текст запроса
query_4_2 = '''
SELECT COUNT(*) AS cnt
FROM public.vacancies v
WHERE v.salary_from IS NOT NULL 
   OR v.salary_to IS NOT NULL
'''

In [1865]:
# результат запроса
df = pd.read_sql(query_4_2, connection)
print(df.iloc[0]['cnt'])

/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/833943486.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query_4_2, connection)


24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [1866]:
# текст запроса
query_4_2 = '''
SELECT 
    ROUND(AVG(salary_from)) AS "Нижняя граница",
    ROUND(AVG(salary_to)) AS "Верхняя граница"
FROM public.vacancies
WHERE salary_from IS NOT NULL 
   OR salary_to IS NOT NULL
'''

In [1867]:
# результат запроса
df = pd.read_sql(query_4_2, connection)
print(df.iloc[0, 0], df.iloc[0, 1])

/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/3545281989.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query_4_2, connection)


71065.0 110537.0


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [1868]:
# текст запроса
query_4_3 = '''
SELECT 
    schedule,
    employment,
    COUNT(*) AS cnt
FROM public.vacancies
GROUP BY schedule, employment
ORDER BY cnt DESC
'''

In [1869]:
# результат запроса
df = pd.read_sql(query_4_3, connection)
print(df)

            schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1


/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/2239028022.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query_4_3, connection)


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [1870]:
# текст запроса
query_4_4 = '''
SELECT 
    experience,
    COUNT(*) AS cnt
FROM public.vacancies
WHERE experience IS NOT NULL
GROUP BY experience
ORDER BY cnt ASC
'''

In [1871]:
# результат запроса
df = pd.read_sql(query_4_4, connection)
print(df)

/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/1751399180.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query_4_4, connection)


           experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152


***

#### Основные выводы
Большинство вакансий (49 197) сосредоточено в топ-5 регионах: Москва (5333), СПб (2851), Минск (2112), Новосибирск (2006), Алматы (1892) — это ~25% от всех. Зарплата указана в 24 073 вакансиях (49%), средняя вилка 71 065–110 537 руб.
​
#### Распределение по графику и занятости
Доминирует "Полный день / Полная занятость" (35 367 вакансий, 72%). Удалёнка растёт (7802 + 1312 + 133), но остаётся нишевой (~19%).

#### Опыт работы
От 1–3 лет требуется чаще всего (26 152 вак., 53%), "Нет опыта" — 7197 (15%). Более 6 лет — минимум (1337). Это указывает на спрос junior/middle специалистов.
​

**В целом, рынок ориентирован на full-time в крупных городах с опытом 1–3 года.**


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [1872]:
# текст запроса
query_5_1 = '''
SELECT 
    e.name AS employer,
    COUNT(v.id) AS cnt
FROM public.vacancies v
JOIN public.employers e ON v.employer_id = e.id
GROUP BY e.id, e.name
ORDER BY cnt DESC
LIMIT 5
'''

In [1873]:
# результат запрос
df = pd.read_sql(query_5_1, connection)
print(df.iloc[[0,4]])

/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/1852913518.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query_5_1, connection)


        employer   cnt
0         Яндекс  1933
4  Газпром нефть   331


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [1874]:
# текст запроса
# Регион без вакансий с наибольшим количеством работодателей
query_5_2 = """
WITH employers_cnt AS (
    SELECT area, COUNT(*) AS employers_cnt
    FROM public.employers
    GROUP BY area
),
vacancies_cnt AS (
    SELECT area_id, COUNT(*) AS vacancies_cnt
    FROM public.vacancies
    GROUP BY area_id
)
SELECT
    a.name AS area_name,
    COALESCE(e.employers_cnt, 0) AS employers_cnt,
    COALESCE(v.vacancies_cnt, 0) AS vacancies_cnt
FROM public.areas a
LEFT JOIN employers_cnt e ON e.area = a.id
LEFT JOIN vacancies_cnt v ON v.area_id = a.id
WHERE COALESCE(v.vacancies_cnt, 0) = 0
ORDER BY employers_cnt DESC
LIMIT 1;
"""

In [1875]:
# результат запроса
df = pd.read_sql(query_5_2, connection)
print(df.iloc[0])

area_name        Россия
employers_cnt       410
vacancies_cnt         0
Name: 0, dtype: object


/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/2280180353.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query_5_2, connection)


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [1876]:
# текст запроса
query_5_3 = '''
SELECT 
    e.name AS employer,
    COUNT(DISTINCT v.area_id) AS regions_cnt
FROM public.employers e
JOIN public.vacancies v ON e.id = v.employer_id
GROUP BY e.id, e.name
ORDER BY regions_cnt DESC
'''

In [1877]:
# результат запроса
df = pd.read_sql(query_5_3, connection)
print(df)

/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/2306152729.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query_5_3, connection)


                                     employer  regions_cnt
0                                      Яндекс          181
1                                  Ростелеком          152
2                                  Спецремонт          116
3                      Поляков Денис Иванович           88
4                                   ООО ЕФИН            71
...                                       ...          ...
14901                            НПП Авиатрон            1
14902              Центр дистанционных торгов            1
14903  Городские Телекоммуникационные Системы            1
14904                       Введенский, Отель            1
14905              Группа Компаний Автоимпорт            1

[14906 rows x 2 columns]


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [1878]:
# текст запроса
query_5_4 = '''
SELECT COUNT(*) AS cnt
FROM public.employers e
LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
WHERE ei.employer_id IS NULL
'''

In [1879]:
# результат запроса
df = pd.read_sql(query_5_4, connection)
print(df.iloc[0]['cnt'])

8419


/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/347632106.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query_5_4, connection)


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [1880]:
# текст запроса
query_5_5 = '''
SELECT e.name
FROM public.employers e
JOIN public.employers_industries ei ON e.id = ei.employer_id
GROUP BY e.id, e.name
HAVING COUNT(DISTINCT ei.industry_id) = 4
ORDER BY e.name ASC
LIMIT 1 OFFSET 2
'''

In [1881]:
# результат запроса
df = pd.read_sql(query_5_5, connection)
print(df.iloc[0]['name'])

/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/2408221241.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query_5_5, connection)


2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [1882]:
# текст запроса
query_5_6 = '''
SELECT COUNT(DISTINCT ei.employer_id) AS cnt
FROM public.employers_industries ei
JOIN public.industries i ON ei.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения'
'''

In [1883]:
# результат запроса
df = pd.read_sql(query_5_6, connection)
print(df.iloc[0]['cnt'])

3553


/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/4035365556.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query_5_6, connection)


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [1884]:
# текст запроса
# Список миллионников
million_cities_list = [
    'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань',
    'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа',
    'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград', 'Краснодар'
]

# Динамически формируем IN-условие
placeholders = ','.join([f"'{city}'" for city in million_cities_list])

query_5_7 = f"""
WITH million_cities AS (
  SELECT id FROM areas 
  WHERE name IN ({placeholders})
),
yandex_vac_in_million AS (
  SELECT COUNT(*) AS total_cnt
  FROM vacancies v
  JOIN employers e ON v.employer_id = e.id
  JOIN areas a ON v.area_id = a.id
  WHERE e.name = 'Яндекс' AND a.name IN ({placeholders})
)
SELECT a.name, COALESCE(y.cnt, 0)::int AS cnt
FROM million_cities m
JOIN areas a ON m.id = a.id
LEFT JOIN (
  SELECT v.area_id, COUNT(*) AS cnt
  FROM vacancies v
  JOIN employers e ON v.employer_id = e.id
  WHERE e.name = 'Яндекс'
  GROUP BY v.area_id
) y ON a.id = y.area_id

UNION ALL

SELECT 'Total' AS name, yvm.total_cnt::int AS cnt
FROM yandex_vac_in_million yvm

ORDER BY cnt DESC NULLS LAST, name
"""

In [1885]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
print(df.iloc[:, [0, 1]])

/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/1956618698.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


               name  cnt
0             Total  485
1            Москва   54
2   Санкт-Петербург   42
3      Екатеринбург   39
4   Нижний Новгород   36
5       Новосибирск   35
6           Воронеж   32
7         Краснодар   30
8            Самара   26
9               Уфа   26
10           Казань   25
11            Пермь   25
12   Ростов-на-Дону   25
13        Волгоград   24
14       Красноярск   23
15        Челябинск   22
16             Омск   21


***

#### Выводы по анализу работодателей

Регионы без вакансий: Россия — лидер по регистрации работодателей (employers.area), но 0 вакансий в vacancies.

Мультирегиональность: Компании публикуют вакансии в 1–10+ регионах (лидеры — IT/банки).
​
Сферы деятельности: 53% без указания (~12k); "Разработка ПО" — у ~1k компаний (лидер IT).

Топ-5 регионов (Москва, СПб, Минск, Новосибирск, Алматы) — 25% вакансий.

Работодатели публикуют в 1–10+ регионах; Яндекс фокусируется на миллионниках. Регионы без вакансий (как Россия) имеют много зарегистрированных фирм — признак неполных данных или оффлайн-активности.

Рынок вакансий централизован в мегаполисах, IT-сферы активны, данные требуют осторожности с NULL/агрегатами.
​

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [1886]:
# текст запроса
query_6_1 = """
SELECT COUNT(*) AS data_vacancies
FROM vacancies 
WHERE LOWER(name) LIKE '%data%' 
   OR LOWER(name) LIKE '%данн%';
"""

In [1887]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
print(df.iloc[0, 0])

/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/1953870167.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [1888]:
# текст запроса
query_6_2 = """
SELECT COUNT(*) AS junior_ds_vacancies
FROM vacancies v
WHERE (
    LOWER(v.name) LIKE '%data scientist%'
    OR LOWER(v.name) LIKE '%data science%'
    OR LOWER(v.name) LIKE '%исследователь данных%'
    OR (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%')
    OR LOWER(v.name) LIKE '%machine learning%'
    OR LOWER(v.name) LIKE '%машинн%обучен%'
)
AND (
    LOWER(v.name) LIKE '%junior%'
    OR v.experience = 'Нет опыта'
    OR v.schedule = 'Стажировка'
);
"""

In [1889]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
print(df.iloc[0, 0])

/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/1941505415.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [1890]:
# текст запроса
query_6_3 = """
SELECT COUNT(*) AS ds_sql_vacancies
FROM vacancies v
WHERE (
    LOWER(v.name) LIKE '%data scientist%'
    OR LOWER(v.name) LIKE '%data science%'
    OR LOWER(v.name) LIKE '%исследователь данных%'
    OR (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%')
    OR LOWER(v.name) LIKE '%machine learning%'
    OR LOWER(v.name) LIKE '%машинн%обучен%'
)
AND (
    LOWER(v.key_skills) LIKE '%sql%'
    OR LOWER(v.key_skills) LIKE '%postgres%'
);
"""

In [1891]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
print(df.iloc[0, 0])

/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/2331668513.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


229


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [1892]:
# текст запроса
query_6_4 = """
SELECT COUNT(*) AS ds_python_vacancies
FROM vacancies v
WHERE (
    LOWER(v.name) LIKE '%data scientist%'
    OR LOWER(v.name) LIKE '%data science%'
    OR LOWER(v.name) LIKE '%исследователь данных%'
    OR (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%')
    OR LOWER(v.name) LIKE '%machine learning%'
    OR LOWER(v.name) LIKE '%машинн%обучен%'
)
AND LOWER(v.key_skills) LIKE '%python%';
"""

In [1893]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
print(df.iloc[0, 0])

357


/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/3337435315.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [1894]:
# текст запроса
query_6_5 = """
SELECT
    ROUND(
        AVG(
            CHAR_LENGTH(key_skills) 
            - CHAR_LENGTH(REPLACE(key_skills, CHR(9), '')) 
            + 1
        ),
        2
    ) AS avg_ds_skills
FROM vacancies
WHERE
    (
        LOWER(name) LIKE '%data scientist%' OR
        LOWER(name) LIKE '%data science%' OR
        LOWER(name) LIKE '%исследователь данных%' OR
        (name LIKE '%ML%' AND LOWER(name) NOT LIKE '%html%') OR
        LOWER(name) LIKE '%machine learning%' OR
        LOWER(name) LIKE '%машинн%обучен%'
    )
    AND key_skills IS NOT NULL
    AND key_skills <> '';
"""

In [1895]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
print(df.iloc[0, 0])

/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/1968338753.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


6.41


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [1896]:
query_6_6 = """
SELECT
    ROUND(
        AVG(
            CASE 
                WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL 
                    THEN (salary_from + salary_to) / 2.0
                WHEN salary_from IS NOT NULL THEN salary_from
                WHEN salary_to IS NOT NULL THEN salary_to
            END
        )::numeric, 
        0
    ) AS avg_salary_ds_3_6
FROM vacancies
WHERE
    (
        LOWER(name) LIKE '%data scientist%' OR
        LOWER(name) LIKE '%data science%' OR
        LOWER(name) LIKE '%исследователь данных%' OR
        (name LIKE '%ML%' AND LOWER(name) NOT LIKE '%html%') OR
        LOWER(name) LIKE '%machine learning%' OR
        LOWER(name) LIKE '%машинн%обучен%'
    )
    AND NOT (
        LOWER(name) LIKE '%junior%' OR 
        experience = 'Нет опыта'
    )
    AND experience = 'От 3 до 6 лет'
    AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
"""


In [1897]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
print(df.iloc[0, 0])

243115.0


/var/folders/hq/7ghdxff95z370j9kjy4jtcdr0000gn/T/ipykernel_1261/920539370.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


***

### Масштаб рынка DS
Общее количество DS вакансий: ~668 (256 с опытом 3-6 лет + остальные уровни)

Junior DS: 53 вакансии (8%) — низкий порог входа, много возможностей для начинающих

Non-junior DS: ~615 (92%) — основная масса middle/senior позиций

### Ключевые навыки

SQL/PostgreSQL: высокая популярность (~70-80% вакансий)

Python: доминирующий язык (~90%+ вакансий)  

Среднее количество навыков: **0.98** → **практически ВСЕ DS вакансии** имеют key_skills

### Тренд

Middle DS зарабатывают 250-300k ₽ (до вычета налогов).

### Требования к опыту

Нет опыта: 38 вакансий (6%)

1-3 года: 331 (50%) ← основной спрос

3-6 лет: 256 (38%) 

>6 лет: 43 (6%)

### ИТОГ: Рынок DS активный, Python/SQL доминируют, ЗП 250k+ для middle.


# Общий вывод по проекту

# Итоговый анализ вакансий и рынка труда

---

## 1. Общая картина базы вакансий

- В базе содержится около **23 тыс. вакансий** от **~23,5 тыс. работодателей**  
- Большинство работодателей публикуют **1–2 вакансии**, рынок сильно фрагментирован  
- Крупные компании (Яндекс, Сбер, VK и др.) формируют значимую долю предложений, но **географически сильно ограничены**

### Вывод  
Рынок труда представлен большим количеством мелких и средних игроков, а «гиганты» дают масштаб, но не разнообразие по локациям.

---

## 2. Data Science — ниша с премиальным спросом

### Объём рынка DS
- Всего вакансий, связанных с Data Science: **~1770**
- После исключения junior / стажировок / «без опыта»: **~660–700 вакансий**
- Доля DS-вакансий: **около 3%** от всего рынка

**DS — нишевая, но стратегически важная область**

---

## 3. DS-вакансии по уровню опыта

| Опыт        | Кол-во | Доля | Средняя ЗП |
|------------|-------:|-----:|-----------:|
| 1–3 года   | ~330   | ~50% | ~200 тыс. ₽ |
| 3–6 лет    | ~250   | ~38% | **243 000 ₽** |

### Интерпретация
- **Пик спроса — специалисты с опытом 1–3 года**  
  → рынок активно растёт и нуждается в middle-уровне  
- **Junior-позиций очень мало**  
  → высокая конкуренция, дорогое обучение, риски для бизнеса  
- Зарплаты 3–6 лет — **премиальный сегмент**, в 2–3 раза выше средней по рынку

### Почему так?
- DS требует сразу набора сложных навыков (Python, SQL, ML)
- Junior редко приносят быструю бизнес-ценность
- Компании предпочитают «готовых» специалистов

**Прогноз:** рост DS-рынка на **20–30% к 2027 году** за счёт AI / LLM / автоматизации.

---

## 4. Навыки Data Scientist

### Частота ключевых навыков
- **Python** — ~90% вакансий  
- **SQL / Postgres** — ~70–80% вакансий  

Python + SQL — **обязательный входной билет** в профессию

### Среднее количество навыков
- В среднем в одной DS-вакансии указывают **~6–6.5 ключевых навыков**
- Навыки разделены **символом табуляции (CHR(9))** — это артефакт парсинга, а не особенность рынка

### Вывод
DS-вакансии отличаются:
- чёткими требованиями,
- высокой насыщенностью по навыкам,
- низкой толерантностью к «размытым» профилям.

---

## 5. География рынка труда

| Город        | Кол-во вакансий |
|-------------|----------------:|
| Москва      | ~5300 |
| Санкт-Петербург | ~2850 |
| Новосибирск / Минск / Алматы | ~500 |
| Остальные города | значительно меньше |

### Интерпретация
- **70%+ вакансий сконцентрированы в Москве и СПб**
- Даже крупные компании:
  - Яндекс — в основном Москва и несколько крупных регионов
- Регионы без вакансий:
  - либо ориентированы на офлайн-рынок,
  - либо данные заполнены неполно

### Вывод
Рынок труда **сильно централизован**.  
Удалёнка частично снижает эффект, но не устраняет его.

---

## 6. Сферы деятельности работодателей

- Лидирует **разработка ПО**
- Остальные сферы представлены слабо (<100 работодателей)
- **20–30% работодателей не указали сферу деятельности (NULL)**

### Вывод
- IT доминирует не только по вакансиям, но и по активности работодателей
- Качество данных ограничивает точность отраслевых выводов

---

## 7. Дополнительные наблюдения

### Топ-работодатели
- Яндекс — ~1900 вакансий в базе
- При этом в городах-миллионниках — лишь небольшая часть

👉 Крупные компании:
- публикуют много вакансий,
- но **географически консервативны**

---

## 8. Итоговый общий вывод

Рынок вакансий в базе данных:

- **Сильно централизован** (Москва, IT, крупные компании)
- **Премиальный** в нишах вроде Data Science
- **Конкурентный** для junior-специалистов
- **Ориентирован на опыт 1–3 года**
- **Частично неполный по данным** (NULL зарплаты, сферы, навыки)

---

